### PR TITLE
SQL: Optimisation fixes for conjunction merges

### PR DESCRIPTION
**Pull request by bpintea**
This PR fixes the following issues around the way comparisions are
merged with ranges in conjunctions:
* the decision to include the equality of the lower limit is corrected;
* the selection of the upper limit is corrected to use the upper bound
of the range;
* the list of terms in the conjunction is sorted to have the ranges at
the bottom; this allows subsequent binary comarisions to find compatible
ranges and potentially be merged away. The end guarantee being that the
optimisation takes place irrespective of the order of the conjunction
terms in the statement.

Some comments are also corrected.

Addresses #49637

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 1]
```diff
@@ -1316,7 +1331,7 @@ private boolean findConjunctiveComparisonInRange(BinaryComparison main, List<Ran
                                     ranges.remove(i);
                                     ranges.add(i,
                                             new Range(other.source(), other.value(),
-                                                    main.right(), lowerEq ? true : other.includeLower(),
+                                                    main.right(), lowerEq ? false : main instanceof GreaterThanOrEqual,
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: That's a big miss - were there no tests for this?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 2]
```diff
@@ -1325,19 +1340,19 @@ private boolean findConjunctiveComparisonInRange(BinaryComparison main, List<Ran
                             }
                         }
                     } else if (main instanceof LessThan || main instanceof LessThanOrEqual) {
-                        if (other.lower().foldable()) {
-                            Integer comp = BinaryComparison.compare(value, other.lower().fold());
+                        if (other.upper().foldable()) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Likewise - a whole optimization being skipped...

**Review by costin — APPROVED** [review]
LGTM.
Thanks for catching the bug while at it - can you please double check all cases are covered by the existing test cases?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java (hunk 1 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: It seemed much like a c&p fallout.
There wasn't a test, no, so I've simply implemented the test listed as comment in the source (`2 < a AND (2 <= a < 3) -> 2 < a < 3` / `testCombineBinaryComparisonsAndRangeLower()`)

**Comment by bpintea** [comment]
> Thanks for catching the bug while at it - can you please double check all cases are covered by the existing test cases?

I hope the coverage should be good now.
There are however some further tunings possible. For instance smth like `(1 < a < 3) OR (2 < a < 4) -> (1 < a < 4)` isn't optimised, or `(1 < a < 2) AND (3 < a < 4)` results into `3 < a < 2`, which will eventually still yield the correct result, but it isn't optimised into a `FALSE` right away. So I'd extend this optimisation, but as a followup PR.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 3]
```diff
@@ -1090,7 +1091,21 @@ private Expression combine(And and) {
 
             boolean changed = false;
 
-            for (Expression ex : Predicates.splitAnd(and)) {
+            List<Expression> andExps = Predicates.splitAnd(and);
+            // Ranges need to show up before BinaryComparisons in list, to allow the latter be optimized away into a Range, if possible
+            andExps.sort(new Comparator<Expression>() {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Maybe extract this custom Comparator into its own variable inside `CombineBinaryComparisons` and re-use that, without creating it each time the `combine` method is called?

**Review by astefan — APPROVED** [review]
Left one comment. Otherwise, LGTM.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java (hunk 3 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Good point, thanks.
I've however replaced with a (non-capturing) lambda, which should then be optimised by the jvm.

**Comment by costin** [comment]
`(1 < a < 3) OR (2 < a < 4) -> (1 < a < 4)`  that would incorrect though - take `3.5` - it doesn't pass the initial condition but it does the latter.

`(1 < a < 2) AND (3 < a < 4) results into 3 < a < 2`
should be optimized on the second run. The idea is that each rule modifies on piece, subsequent optimizations are applied by going through the set of rules multiple times until none match.

**Comment by bpintea** [comment]
> `(1 < a < 3) OR (2 < a < 4) -> (1 < a < 4)` that would incorrect though - take `3.5` - it doesn't pass the initial condition but it does the latter.

For conjunction, that wouldn't work, true, but overlapping disjunctions can be optimised with a simple union (`3.5` satisfies the 2nd range of the disjunction).

> (1 < a < 2) AND (3 < a < 4) results into 3 < a < 2
should be optimized on the second run. The idea is that each rule modifies on piece, subsequent optimizations are applied by going through the set of rules multiple times until none match.

Sure, sure. I was only wondering if this couldn't be achieved in one go.

Anyways, future potential optimisations.

**Comment by costin** [comment]
> For conjunction, that wouldn't work, true, but overlapping disjunctions can be optimised with a simple union (3.5 satisfies the 2nd range of the disjunction).

You're right, we should add that too. If you don't plan on working on it in the near future, please make an issue so we don't lose track of it.

Thanks.

**Pull request merged**